### PR TITLE
Ensure Trillian Log core is ok with adding duplicate entries.

### PR DESCRIPTION
**Pull request by AlCutter**
Goes some way to enforcing #619 

Commits add an in-memory based log storage, then add an integration test which uses it to store and retrieve dupe leaves.

This PR brought to you by the numbers 1 through 12 https://www.youtube.com/watch?v=VOaZbaPzdsk

### CODE REVIEW

File: integration/log.go [hunk 1]
```diff
@@ -119,7 +124,7 @@ func RunLogIntegration(client trillian.TrillianLogClient, params TestParameters)
 
 	// Step 3 - Use get entries to read back what was written, check leaves are correct
 	glog.Infof("Reading back leaves from log ...")
-	leafMap, err := readbackLogEntries(params.treeID, client, params)
+	leafMap, err := readbackLogEntries(params.treeID, client, params, leafCounts)
```
Comment (Martin2112) [inline integration/log.go]: Does this work if you don't queue anything and leafCounts is empty?

File: storage/memory/log_storage.go [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+
+	ret := make([]int64, 0, len(t.ms.trees))
+	for k := range t.ms.trees {
+		ret = append(ret, k)
+	}
+	return ret, nil
+}
+
+func (t *readOnlyLogTX) GetActiveLogIDsWithPendingWork(ctx context.Context) ([]int64, error) {
+	// just retusn all trees for now
```
Comment (Martin2112) [inline storage/memory/log_storage.go]: typo

File: storage/memory/log_storage.go [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+	for _, l := range leaves {
+		q.PushBack(l)
+	}
+	return []*trillian.LogLeaf{}, nil
+}
+
+func (t *logTreeTX) GetSequencedLeafCount(ctx context.Context) (int64, error) {
+	var sequencedLeafCount int64
+
+	t.tx.DescendRange(seqLeafKey(t.treeID, 9223372036854775807), seqLeafKey(t.treeID, 0), func(i btree.Item) bool {
```
Comment (Martin2112) [inline storage/memory/log_storage.go]: large magic number should be documented and perhaps made a const.

File: log/sequencer.go [hunk 4]
```diff
@@ -194,6 +194,13 @@ func (s Sequencer) SequenceBatch(ctx context.Context, logID int64, limit int) (i
 	// TODO(mhs): Might be better to create empty root in provisioning API when it exists
 	if currentRoot.RootHash == nil {
 		glog.Warningf("%v: Fresh log - no previous TreeHeads exist.", logID)
+		// SignRoot starts a new transaction, and we've gone one open here until
```
Comment (AlCutter) [inline log/sequencer.go]: done

File: storage/memory/admin_storage.go [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+	return s.Begin(ctx)
+}
+
+func (s *memoryAdminStorage) Begin(ctx context.Context) (storage.AdminTX, error) {
+	return &adminTX{ms: s.ms}, nil
+}
+
+type adminTX struct {
+	ms *memoryTreeStorage
+	// mu guards *direct* reads/writes on closed, which happen only on
```
Comment (AlCutter) [inline storage/memory/admin_storage.go]: This comment came in with the mysql fork, have you got a suggestion for a better wording?

File: log/sequencer.go [hunk 6]
```diff
@@ -194,6 +194,13 @@ func (s Sequencer) SequenceBatch(ctx context.Context, logID int64, limit int) (i
...
+		// this function returns.
+		// This explicit Close() is a work-around for the in-memory storage which
+		// locks the tree for each TX.
+		// TODO(al): The Close() is benign enough since implementations are all
+		// idempotent, but would be nice to not have to worry about it.
```
Comment (AlCutter) [inline log/sequencer.go]: SGTM, but I'd prefer that's done in another PR, I've updated the TODO.

File: storage/memory/doc.go [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+// Package memory provides a simple in-process implementation of the tree- and
+// log-storage interfaces.
+//
+// This implementation is intended SOLELY for use in integration tests which
+// exercise properties of the higher levels of Trillian componened - e.g.
+// an integration test which ensures that the Trillian Log is able to correctly
+// handle a tree which contains duplicate leaves.
+//
+// The storage implementation is based on a BTree, which provide an ordered
```
Comment (AlCutter) [inline storage/memory/doc.go]: done

File: storage/memory/admin_storage.go (hunk 5 above)
Comment (AlCutter) [inline storage/memory/admin_storage.go]: removed "*direct*" after chatting IRL

File: storage/memory/doc.go [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+// key-value space which can be used to store arbitrary items, as well as
+// scan ranges of keys in order.
+//
+// The implementation does provide transaction-like semantics for the
+// LogStorage interface, although conflict is avoided by each writa-transaction
```
Comment (AlCutter) [inline storage/memory/doc.go]: *Writa-transaction*  I think it could really catch on...
*cough* done.

File: integration/log.go (hunk 1 above)
Comment (AlCutter) [inline integration/log.go]: Hmm, I think it'd work, but that would be a rubbish integration test :)

File: storage/memory/log_storage.go (hunk 2 above)
Comment (AlCutter) [inline storage/memory/log_storage.go]: >_< done

File: log/sequencer.go (hunk 6 above)
Comment (Martin2112) [inline log/sequencer.go]: Ack.

File: storage/memory/log_storage.go (hunk 3 above)
Comment (AlCutter) [inline storage/memory/log_storage.go]: dones

File: integration/log.go (hunk 1 above)
Comment (Martin2112) [inline integration/log.go]: Haha, no I meant if you run the test client with the option where it skips queuing the map will be empty and it gets used later. Possibly it still works, not sure. It's ok for the test runs as they never set this.

File: storage/memory/tree_storage.go [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+	ret.store.ReplaceOrInsert(k)
+
+	k = hashToSeqKey(t.TreeId)
+	k.(*kv).v = make(map[string][]int64)
+	ret.store.ReplaceOrInsert(k)
+
+	return ret
+}
+
+func (m *memoryTreeStorage) beginTreeTX(ctx context.Context, readonly bool, treeID int64, hashSizeBytes int, strataDepths []int, populate storage.PopulateSubtreeFunc, prepare storage.PrepareSubtreeWriteFunc) (treeTX, error) {
```
Comment (AlCutter) [inline storage/memory/tree_storage.go]: Passed in a subtreeCache directly rather than all the bits necessary to build one [also from the mysql fork ;)]

Comment (Martin2112) [inline storage/memory/tree_storage.go]: Have you ever heard the phrase "no blame culture"? I'll look at that code later as it would help there.

Comment (AlCutter) [inline storage/memory/tree_storage.go]: hehe :)
